packages/flutter_test/lib/src/widget_tester.dart

```diff
@@ -609,11 +609,60 @@ class WidgetTester extends WidgetController implements HitTestDispatcher, Ticker
   /// this method again. Attempts to do otherwise will result in a
   /// [TestFailure] error being thrown.
+  ///
+  /// If your test code depends on the result of a task that did not complete,
+  /// the most common symptom would be a hanging test as FakeAsync environment
+  /// is unable to resolve a future that was created in `runAsync`. If you
+  /// observe such behavior or flakiness, you have a number of options:
+  ///
+  /// - Consider restructuring your code so you do not need `runAsync`. This is
+  ///   the optimal solution as widget tests are designed to run in Fake async
+  ///   environment.
+  ///
+  /// - Expose a [Future] in your application code that signals the readiness of
+  ///   your widget tree, then await that future inside `callback`.
+  ///
+  /// - If either option is not feasible, await [resolveFutures] inside
+  ///   `callback`.
   Future<T> runAsync<T>(
     Future<T> callback(), {
     Duration additionalTime = const Duration(milliseconds: 1000),
   }) => binding.runAsync<T>(callback, additionalTime: additionalTime);
 
+  /// Waits for a stable state within a single `runAsync` execution.
+  ///
+  /// Note that using this method is *not* recommended. This method should only
+  /// be used when there are unawait-able side effects that your test depends
+  /// on.
+  ///
+  /// [runAsync] method keeps track of microtasks and timers created during
+  /// `callback`. [resolveFutures] creates a periodic timer that checks the
+  /// status of these timers and returns once there are no active timers *and*
+  /// no new microtasks have been created for a certain period of time. In other
+  /// words, it attempts to observe a certain length of "idle" period specified
+  /// by `expectedIdlePeriod`.
+  ///
+  /// Microtasks might still be generated after this idle period, therefore the
+  /// completion of this method does not mean the application is in a stable
+  /// state. It also means each test that awaits this Future extends its runtime
+  /// by this idle period since we have to wait *at least* that length of time
+  /// to observe the idleness. This is why [runAsync] does not await on
+  /// `resolveFutures` by default.
+  ///
+  /// Setting `expectedIdlePeriod` is a tradeoff between how much longer your
+  /// test will take and how confident we are of the idle state. If you set it
+  /// too long, you will waste a lot of test time. If you set it too short, your
+  /// tests might be flaky. See other recommended approaches for dealing with
+  /// unawaited side effects in [runAsync] documentation.
+  ///
+  /// Note that if you are creating a new isolate during the `runAsync`
+  /// execution (for instance by using [compute]), you need to await its result.
+  /// If you don't, `runAsync` cannot track async operations created in this
+  /// isolate.
+  Future<void> resolveFutures({
+    Duration expectedIdlePeriod = const Duration(milliseconds: 500),
```

Comment:
I think this should default to infinite. For example, how should we pick a right time for this? Either a duration that is super big or a duration that is close to the actual run time. The later will make test flaky.

Reply:
If this is infinite, this Future will never complete.

This isn't a timeout. It is the length of idle period that we expect to observe before declaring the application as idle.

Suppose we have the following scenario and for simplicity, assume that microtasks are executed as soon as they are scheduled:

```
microtask1 <--29 ms --> microtask2 < -----------------143 ms -------> microtask3
```

If we have `expectedIdlePeriod` as 20ms, this function will return after microtask1. If we set it to 50ms, it will return after microtask2 because it will observe that there are no pending timers or microtasks but the amount of time that passed **since** the latest microtask (microtask1) is less than 50ms.  After waiting for ~50ms after microtask2 executed, it will run and claim that enough time has passed. If you set it to 500ms, all three microtasks will need to execute before the application is declared idle.

The longer you set this, the more confident you will be of idleness. However, it will also lengthen your test execution by that much. This is real-time wait.

Reply:
Ah I see, it only resolve when 
delta_time > expectedIdlePeriod and the timer/microtask caches are empty.
That is a bit weird though, shouldn't we resolve the future as soon as timer/microtask caches are empty? why do we still want to wait in this case.
I was imagining this kind of logic

if (delta_time > expectedIdlePeriod) throws
else if (caches are empty) resolveFuturesCompleter.complete();
else continue;

Reply:
That was the first approach I tried but it does not work if your microtasks complete quickly but are scheduled by external events (such as network or disk io). That is the normal case.  For instance, if I am listing directories and stat'ing them one by one, this is what you would see in the microtask queue:

```
0 milliseconds: start
2 ms: end
15 ms: start
17 ms: end
...
```

For the vast majority of time (e.g. 2 to 15 ms), the queue is empty so you will prematurely think that the program is done.

That's why I had to introduce the idea of "idleness" into the code. Wait for a "silent" period where no microtask is fired. The length of the period is controllable by the user.

Reply:
i understand now. Although we make it more transparent on what is under the hood of async zone, we still have not solve the problem that this api may encourage flakiness. I don't have a good answer neither but I have some other idea that i would like to share.

It is possible to write the resolveFuture to take in a integer number `count`, and the resolveFuturesCompleter will only resolve when there are at least that # of timers has gone from active -> non-active. This way we no longer rely on the duration of the future but the # of the future that should resolve in a test, and that should never change.

Reply:
I would be against that because it would make the tests really difficult to maintain. Suppose that we did a Flutter roll that changed the internals of PictureRecorder (added a new Future). Now all of a sudden we have to update all the tests and increase this number by 1. It is fairly arbitrary and does not add more insight to developer. They just have to guess the right number.

pumpAndSettle is the same by the way. We recommend that people test the number of frames pumped returned by pumpAndSettle() to make sure they got the right number of frames but in reality, no one does that. It is very difficult to maintain and it can change easily by your dependencies.

Reply:
> pumpAndSettle is the same by the way. We recommend that people test the number of frames pumped returned by pumpAndSettle() to make sure they got the right number of frames but in reality, no one does that. It is very difficult to maintain and it can change easily by your dependencies.

pumpAndSettle really only makes sense to test your own widget, it doesn't make much sense when testing something where you don't know how many frames are going to run. In general, app logic shouldn't be triggered on frames, so when testing app logic you shouldn't need to wait for things to settle. Just pump a frame to start animations, and pump a frame with a delay sufficiently later that all the animations are done.